plasmapy/tests/helpers/tests/test_expected.py

```diff
@@ -0,0 +1,178 @@
+    Case(expected_value, "expecting_a_value", True),
+    Case(expected_value_and_warning, "expected_warning", expected_warning),
+    Case(expected_value_and_warning, "expected_value", expected_value),
+    Case(expected_value_and_warning, "expecting_an_exception", False),
+    Case(expected_value_and_warning, "expecting_a_warning", True),
+    Case(expected_value_and_warning, "expecting_a_value", True),
+    Case(expected_value_and_warning, "expected_outcome", expected_warning_and_value),
+    Case(expected_warning_and_value, "expected_warning", expected_warning),
+    Case(expected_warning_and_value, "expected_value", expected_value),
+    Case(expected_warning_and_value, "expecting_an_exception", False),
+    Case(expected_warning_and_value, "expecting_a_warning", True),
+    Case(expected_warning_and_value, "expecting_a_value", True),
+    Case(expected_warning_and_value, "expected_outcome", expected_warning_and_value),
+]
+
+
+@pytest.mark.parametrize("case", expected_outcome_test_cases)
+def test_expected_test_outcome_attributes(case: Case):
+    """Test that `ExpectedTestOutcome` attributes each return the expected values."""
+    expected_outcome = ExpectedTestOutcome(case.argument)
+    result = expected_outcome.__getattribute__(case.attribute)
+    if result is not case.correct_outcome and result != case.correct_outcome:
+        pytest.fail(
+            f"ExpectedTestOutcome({case.argument}) results in {repr(result)} "
+            f"but should result in {case.correct_outcome}."
+        )
+
+
+exception_to_be_raised = RuntimeError
+
+exception_raising_cases = [
+    Case(expected_exception, "expected_warning", exception_to_be_raised),
+    Case(expected_exception, "expected_value", exception_to_be_raised),
+    Case(expected_warning, "expected_exception", exception_to_be_raised),
+    Case(expected_warning, "expected_value", exception_to_be_raised),
+    Case(expected_value, "expected_exception", exception_to_be_raised),
+    Case(expected_value, "expected_warning", exception_to_be_raised),
+    Case(expected_value_and_warning, "expected_exception", exception_to_be_raised),
+    Case(expected_warning_and_value, "expected_exception", exception_to_be_raised),
+]
+
+
+@pytest.mark.parametrize("case", exception_raising_cases)
+def test_expected_test_outcome_exceptions(case: Case):
+    """Test that attributes of `ExpectedTestOutcome` raise exceptions as needed."""
+    if not issubclass(case.correct_outcome, Exception):
+        raise TypeError(
+            "Incorrect test setup: the expected outcome must be an exception."
+        )
+    with pytest.raises(case.correct_outcome):
+        expected_outcome = ExpectedTestOutcome(case.argument)
+        result = expected_outcome.__getattribute__(case.attribute)
+        pytest.fail(
+            f"The ExpectedTestOutcome instance for {case.argument} did not "
+            f"raise the expected exception but instead returned {result}."
+        )
```

Comment:
I'm trying and failing to wrap my head around this.

So this is in a `raises` block, but your error message says it **did not** raise the expected exception.

... wha?

Can we rewrite this to avoid the manual `pytest.fail`?

At the very least,
```suggestion
    with pytest.raises(case.correct_outcome):
        expected_outcome = ExpectedTestOutcome(case.argument)
        result = getattr(expected_outcome, case.attribute)
        pytest.fail(
            f"The ExpectedTestOutcome instance for {case.argument} did not "
            f"raise the expected exception but instead returned {result}."
        )
```